src/charm.py

```diff
@@ -1237,45 +865,48 @@ def _provider_from_relation(self, relation: Relation):
         raise RuntimeError(f"Invalid relation type: {relation_type} ({relation.name})")
 
     @property
-    def external_host(self):
+    def _external_host(self) -> Optional[str]:
         """Determine the external address for the ingress gateway.
 
         It will prefer the `external-hostname` config if that is set, otherwise
         it will look up the load balancer address for the ingress gateway.
 
         If the gateway isn't available or doesn't have a load balancer address yet,
-        returns None.
+        returns None. Only use this directly when external_host is allowed to be None.
         """
         if external_hostname := self.model.config.get("external_hostname"):
             return external_hostname
 
         return _get_loadbalancer_status(namespace=self.model.name, service_name=self.app.name)
 
     @property
-    def _routing_mode(self) -> _RoutingMode:
+    def external_host(self) -> str:
```

Comment:
we need external_host in two kinds of situations: in one of them it's allowed to be None, in the other it's not and we raise an exception. This division makes it explicit.
When you allow a None and don't care: use the private _external_host, when you need it to be a string and want to raise otherwise, use the public variant.

Reply:
It does, and we do: we have `external_host` that is guaranteed to not raise, but returns Optional[str]. That allows you to code defensively if that's your jam.
All I'm proposing is we **add** api for the other case, the one in which we should raise if external host is not set (because coding defensively is only nice at the toplevel logic layer, you don't want to keep questioning yourself when you're ten levels deep).

Reply:
@sed-i is this a blocker we need to discuss more in depth about, or shall I consider you convinced and on-board?